File: services/groupbuy-service/src/main/java/groupbuy_service/participation/event/OrderCancelledEventListener.java
package groupbuy_service.participation.event;

import groupbuy_service.global.reconciliation.service.DltReconciliationService;
import groupbuy_service.order.event.OrderCancelledEvent;
import groupbuy_service.participation.service.ParticipationService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.kafka.annotation.BackOff;
import org.springframework.kafka.annotation.DltHandler;
import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.kafka.annotation.RetryableTopic;
import org.springframework.kafka.retrytopic.DltStrategy;
import org.springframework.kafka.support.KafkaHeaders;
import org.springframework.messaging.handler.annotation.Header;
import org.springframework.stereotype.Component;
import tools.jackson.core.JacksonException;

@Slf4j
@Component
@RequiredArgsConstructor
public class OrderCancelledEventListener {

    private final ParticipationService participationService;
    private final DltReconciliationService dltReconciliationService;

    @RetryableTopic(
            attempts = "3",
            backOff = @BackOff(delay = 1000, multiplier = 2.0, maxDelay = 20000),
            exclude = {JacksonException.class},
            dltStrategy = DltStrategy.FAIL_ON_ERROR,
            autoCreateTopics = "true"
    )
    @KafkaListener(topics = OrderCancelledEvent.TOPIC, groupId = "groupbuy-service-group")
    public void onOrderCancelledEvent(OrderCancelledEvent event) throws Exception {
        log.info("[groupbuy-service] 주문 취소 이벤트 수신: participationId={}", event.participationId());
        participationService.failParticipation(event.participationId());
    }

    @DltHandler
    public void handleDlt(
            OrderCancelledEvent event,
            @Header(KafkaHeaders.DLT_ORIGINAL_TOPIC) String topic,
            @Header(value = KafkaHeaders.DLT_EXCEPTION_MESSAGE, required = false) String exceptionMessage
    ) {
        String errorMessage = exceptionMessage != null
                ? exceptionMessage
                : "FailedEvent: OrderCancelledEvent";

        log.error("🚨 [DLT] 주문 취소 처리 최종 실패. topic: {}, event: {}, content: {}", topic, event, errorMessage);
        dltReconciliationService.logFailedEvent(topic, event, errorMessage);
    }
}

File: services/inventory-service/src/main/kotlin/inventory_service/event/OrderCancelledEventListener.kt
package inventory_service.event

import inventory_service.global.reconciliation.DltReconciliationService
import inventory_service.service.InventoryService
import io.github.oshai.kotlinlogging.KotlinLogging
import org.springframework.kafka.annotation.BackOff
import org.springframework.kafka.annotation.DltHandler
import org.springframework.kafka.annotation.KafkaListener
import org.springframework.kafka.annotation.RetryableTopic
import org.springframework.kafka.retrytopic.DltStrategy
import org.springframework.kafka.support.KafkaHeaders
import org.springframework.messaging.handler.annotation.Header
import org.springframework.stereotype.Component
import tools.jackson.core.JacksonException

@Component
class OrderCancelledEventListener(
    private val inventoryService: InventoryService,
    private val dltReconciliationService: DltReconciliationService
) {
    private val log = KotlinLogging.logger {}

    @RetryableTopic(
        attempts = "3",
        backOff = BackOff(delay = 1000, multiplier = 2.0, maxDelay = 20000),
        exclude = [JacksonException::class],
        dltStrategy = DltStrategy.FAIL_ON_ERROR,
        autoCreateTopics = "true"
    )
    @KafkaListener(topics = [OrderCancelledEvent.TOPIC], groupId = "inventory-service-order-cancelled")
    fun restoreStock(event: OrderCancelledEvent) {
        inventoryService.increaseStock(event.productId, event.quantity)
    }

    @DltHandler
    fun handleDlt(
        event: OrderCancelledEvent,
        @Header(value = KafkaHeaders.ORIGINAL_TOPIC, required = false) originalTopic: String?,
        @Header(value = KafkaHeaders.ORIGINAL_PARTITION, required = false) originalPartition: Int?,
        @Header(value = KafkaHeaders.ORIGINAL_OFFSET, required = false) originalOffset: Long?,
        @Header(value = KafkaHeaders.EXCEPTION_MESSAGE, required = false) exceptionMessage: String?
    ) {
        val topic = originalTopic ?: OrderCancelledEvent.TOPIC
        val errorMessage = exceptionMessage ?: "FailedEvent: OrderCancelledEvent"

        log.error { "[DLT] 주문취소 재고 처리 최종 실패. topic=$topic, error=$errorMessage, content=$event" }
        dltReconciliationService.logFailedEvent(topic, originalPartition, originalOffset, event, errorMessage)
    }
}
